### PR TITLE
dialects: Add HostRegisterOp to GPU

### PR DESCRIPTION
**Pull request by PapyChacal**
This small PR just adds the HostRegisterOp to the GPU dialect.
(It makes a host memref accessible from the device/GPU)

### CODE REVIEW

File: xdsl/dialects/gpu.py [hunk 1]
```diff
@@ -240,6 +241,17 @@ def get(dim: DimensionAttr) -> GridDimOp:
                                attributes={"dimension": dim})
 
 
+@irdl_op_definition
+class HostRegisterOp(Operation):
```
Comment (webmiche) [inline xdsl/dialects/gpu.py:245]: At this point I am starting to get towards the limits of my technical GPU knowledge. I am guessing that this hosts a memref to the GPU? 

Maybe this is a good point to point out that many of this operations could use some documentation (the others can be done in a different PR) :)

**Review by webmiche — APPROVED** [review]
LGTM. I am generally starting to feel like your example is somewhat exploding, though maybe this is all boiler plate that a GPU just needs, I really don't know. Maybe consider moving some stuff into smaller test cases (different PR). If you just had a testing framework that you could use that does not look like integration tests, but makes things really deliberately look like testing code. That would be something! :wink:

**Comment by PapyChacal** [comment]
Yeah, example might be a bad name at this point. I'm just throwing every GPU op at it really.
Is the test dialect merged?

File: xdsl/dialects/gpu.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/dialects/gpu.py:245]: It is at my limits too!
This makes the Memref accessible from the device/GPU, and still accessible but the host.
How exactly is deliberately not clear I think, GPU being a high-level dialect to abstract vendors and specific models.

Basically I couldn't document that more than what MLIR did. Should I copy/rephrase it here in docstring?

Comment (webmiche) [reply xdsl/dialects/gpu.py:245]: Yeah, I think there is value in having their docstring here instead of having to look it up.

**Comment by webmiche** [comment]
Not yet. Hopefully soon though...

File: xdsl/dialects/gpu.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/dialects/gpu.py:245]: Sure, makes sense!

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/gpu.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/dialects/gpu.py:245]: @webmiche docstring added, do you have a comment on it?
It's copy-pasted from MLIR's doc directly, so I wouldn't change it too much (as we use their "semantics" ATM), but rephrasing could be fine.

Comment (webmiche) [reply xdsl/dialects/gpu.py:245]: LGTM!

**Review by webmiche — APPROVED** [review]

**Review by AntonLydike — APPROVED** [review]
I'm not deeply familiar with GPU, but looks good to me

File: tests/dialects/test_gpu.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 from typing import List
-from xdsl.dialects import builtin, arith
-from xdsl.dialects.gpu import AllReduceOp, AllReduceOperationAttr, AsyncTokenType, BarrierOp, BlockDimOp, BlockIdOp, GlobalIdOp, GridDimOp, LaneIdOp, LaunchOp, ModuleEndOp, ModuleOp, DimensionAttr, NumSubgroupsOp, SetDefaultDeviceOp, SubgroupIdOp, SubgroupSizeOp, TerminatorOp, ThreadIdOp, YieldOp
+from xdsl.dialects import builtin, arith, memref
+from xdsl.dialects.gpu import AllReduceOp, AllReduceOperationAttr, AsyncTokenType, BarrierOp, BlockDimOp, BlockIdOp, GlobalIdOp, GridDimOp, HostRegisterOp, LaneIdOp, LaunchOp, ModuleEndOp, ModuleOp, DimensionAttr, NumSubgroupsOp, SetDefaultDeviceOp, SubgroupIdOp, SubgroupSizeOp, TerminatorOp, ThreadIdOp, YieldOp
```
Comment (math-fehr) [inline tests/dialects/test_gpu.py:3]: I missed this in the previous PRs, but could you wrap the rhs of the `import` in parentheses?
That makes it wrap on 80 characters.

**Review by math-fehr — APPROVED** [review]
Looks good to me!

File: tests/dialects/test_gpu.py (hunk 2 above)
Comment (PapyChacal) [reply tests/dialects/test_gpu.py]: Sure, done!

**Pull request merged**